File: monitoring-backend/app/schemas/device.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class LibreNMSRegisterRequest(BaseModel):
    hostname: str = Field(..., description="IP or hostname to add into LibreNMS")
    community: str = Field(default="public")
    snmp_version: str = Field(default="v2c")
    port: int = Field(default=161)
    transport: str = Field(default="udp")
    force_add: bool = Field(default=False)

    node_type: Optional[str] = Field(
        default=None, description='Either "device" or "switch"'
    )

    # Optional info for DB record
    name: Optional[str] = None
    device_type: Optional[str] = None
    location_id: Optional[int] = None
    switch_id: Optional[int] = None
    node_id: Optional[int] = None
    description: Optional[str] = None


# Syncing all devices (include switch) from LibreNMS
class AllDevicesSyncConfig(BaseModel):
    default_location_id: int = Field(1, description="Default location for new devices")
    update_existing: bool = Field(True, description="Update already-synced devices")


class AllDevicesSyncReport(BaseModel):
    """Sync result report"""

    created: list[dict]
    updated: list[dict]
    errors: list[dict]
    summary: dict


class DeviceUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    location_id: Optional[int] = Field(None, gt=0)
    device_type: Optional[str] = None
    ip_address: Optional[str] = None
    mac_address: Optional[str] = None
    status: Optional[str] = None
    switch_id: Optional[int] = None
    description: Optional[str] = Field(None, max_length=500)
    librenms_device_id: Optional[int] = None


class DeviceResponse(BaseModel):
    device_id: int
    name: str
    ip_address: str
    mac_address: Optional[str]
    device_type: Optional[str]
    location_id: Optional[int]
    switch_id: Optional[int]
    status: str
    librenms_device_id: Optional[int]
    librenms_hostname: Optional[str]
    librenms_last_synced: Optional[datetime]
    description: Optional[str]
    created_at: datetime
    updated_at: datetime
    last_replaced_at: Optional[datetime]

    class Config:
        from_attributes = True  # Allows conversion from SQLAlchemy model


class BulkLiveDetailsRequest(BaseModel):
    device_ids: List[int]


# Schema for device with location info for map display
class DeviceWithLocation(BaseModel):
    device_id: int
    name: str
    ip_address: str
    mac_address: Optional[str]
    device_type: Optional[str]
    status: str
    description: Optional[str]
    last_replaced_at: Optional[datetime]
    # Location info
    latitude: float
    longitude: float
    location_name: str

File: monitoring-frontend/lib/models/device.dart
class BaseNode {
  final int? id;
  final String name;
  final String ipAddress;
  final String? macAddress;
  final String? deviceType;
  final String nodeKind;
  String? status;
  final String? locationName;
  final int? locationId;
  final int? switchId;
  final int? nodeId;
  final String? description;
  final String? lastReplacedAt;
  final int? librenmsId;

  BaseNode({
    required this.id,
    required this.name,
    required this.ipAddress,
    this.macAddress,
    this.deviceType,
    required this.nodeKind,
    this.status,
    this.locationName,
    this.locationId,
    this.switchId,
    this.nodeId,
    this.description,
    this.lastReplacedAt,
    this.librenmsId,
  });

  factory BaseNode.fromDeviceJson(Map<String, dynamic> json) {
    return BaseNode(
        id: json['device_id'],
        name: json['name'],
        ipAddress: json['ip_address'],
        macAddress: json['mac_address'],
        deviceType: json['device_type'],
        nodeKind: 'device',
        status: json['status'],
        locationId: json['location_id'],
        locationName: json['location_name'],
        switchId: json['switch_id'],
        description: json['description'],
        lastReplacedAt: json['last_replaced_at'],
        librenmsId: json['librenms_device_id']);
  }

  Map<String, dynamic> toDeviceCreateJson() {
    return {
      'name': name,
      'ip_address': ipAddress,
      'mac_address': macAddress,
      'device_type': deviceType,
      'node_kind': nodeKind,
      'status': status,
      'location_id': locationId,
      'switch_id': switchId,
      'description': description
    };
  }

  factory BaseNode.fromSwitchJson(Map<String, dynamic> json) {
    return BaseNode(
        id: json['switch_id'],
        name: json['name'],
        ipAddress: json['ip_address'],
        status: json['status'],
        nodeKind: 'switch',
        deviceType: 'Switch',
        locationName: json['location_name'],
        locationId: json['location_id'],
        nodeId: json['node_id'],
        description: json['description'],
        lastReplacedAt: json['last_replaced_at'],
        librenmsId: json['librenms_device_id']);
  }

  Map<String, dynamic> toSwitchCreateJson() {
    return {
      'name': name,
      'ip_address': ipAddress,
      'status': status,
      'location_id': locationId,
      'node_id': nodeId,
      'node_kind': nodeKind,
      'description': description
    };
  }
}

class NodePage {
  final List<BaseNode> items;
  final int total;
  final int page;
  final int pageSize;

  NodePage({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory NodePage.fromJson(Map<String, dynamic> json) {
    final raw = json['items'] as List? ?? [];
    return NodePage(
      items: raw.map((e) {
        if (e['node_kind'] == 'switch') {
          return BaseNode.fromSwitchJson({
            ...e,
            'switch_id': e['id'],
          });
        }
        return BaseNode.fromDeviceJson({
          ...e,
          'device_id': e['id'],
        });
      }).toList(),
      total: json['total'] ?? 0,
      page: json['page'] ?? 1,
      pageSize: json['page_size'] ?? 10,
    );
  }
}
